### ISSUE TITLE
User dictionary

### ISSUE THREAD

**Issue by bbqsrc** [divvun/giellakbd-ios#47]
Placeholder, to be documented.

**Comment by dylanhand** [comment]
>     * Make sure the algorithm for determining which user dictionary is a separate function so it can be easily modified

This implies there will be multiple user dictionaries. Can you elaborate? I had imagined it would be one dictionary per device with multiple "candidate words" and "confirmed words".

**Comment by bbqsrc** [comment]
It's a typo, fixed.

**Comment by dylanhand** [comment]
- [ ] add 'User Dictionary' button to hosting app UI
- [ ] tapping 'User Dictionary" button loads a table listing all the words in the user dictionary
- [ ] tapping a word in the 'User Dictionary' list shows the contexts in which the word was used
- [ ] add button that allows the user to submit their word dictionary so we can improve our spelling suggestions

**Comment by snomos** [comment]
> > ```
> > * Make sure the algorithm for determining which user dictionary is a separate function so it can be easily modified
> > ```
> 
> 
> This implies there will be multiple user dictionaries. Can you elaborate? I had imagined it would be one dictionary per device with multiple "candidate words" and "confirmed words".

Wouldn't it be natural to have a separate user dictionary for each language/keyboard? You don't want to get suggestions for a Sámi language when typing a completely different language. Maybe the average user don't need this, but at least make sure the code is such that it won't be much work to enable it if we find out this is what people want.

**Comment by dylanhand** [comment]
> Wouldn't it be natural to have a separate user dictionary for each language/keyboard? You don't want to get suggestions for a Sámi language when typing a completely different language. Maybe the average user don't need this, but at least make sure the code is such that it won't be much work to enable it if we find out this is what people want.

Good point. We'll make sure there is a unique dictionary per keyboard locale.

**Comment by snomos** [comment]
>  if user adds word that already exists in the db, do nothing... or, mark it as user word if not already?

What if each time the word is seen, a counter is incremented, and this can then be used to promote it as a suggestion in various contexts (both spelling error and future predictions)

**Comment by bbqsrc** [comment]
That, right now, would go a bit beyond the scope of what we can reasonably do. Let's put that on the backlog to visit once we have the confined case working. Getting context _at all_ is quite a big leap forward, so we can discuss the particular data collection in a separate issue. :D

**Comment by dylanhand** [comment]
Things to clean up:

- [ ] KeyboardLocales/KeyboardLocale
- [ ] Make classes final
- [ ] use class for context instead of struct?
- [ ] take another look at generic settings viewcontrollers - can it be cleaned? protocol? rows lazy so not generated at each access?
- [ ] should `CursorContext` and `WordContext` be combined into one object?
- [ ] update divvunspell to use named parameters in tuple for `offset` and `word`
- [ ] split up UserDictionary so test methods are in an extension

**Comment by snomos** [comment]
I have installed build 155 from this night, but the user dictionary feature is nowhere to be found. I have uninstalled the previous version, restarted the phone, and ensured that full access is given to the keyboards. Could it be that the CI configuration is the one with disabled user dictionary support?

**Comment by dylanhand** [comment]
We’re still testing the CI configuration that enables it. The next couple
of builds may or may not have it - still trying to get the CI right :)
